### PR TITLE
Add FTP server and client connectors

### PR DESCRIPTION
**Pull request by pubudu91**
This PR will add support for FTP Server Connector and FTP Client Connector to Ballerina. Samples demonstrating the usage of these connectors are given below.

The dirURI needs to be of the form: 
`ftp://username:password@ip-address/path/to/the/directory/or/file`

FTP Server Connector sample
```
import ballerina.net.ftp;
import ballerina.lang.system;
import ballerina.lang.messages;

@ftp:config{
	dirURI: "ftp://localhost/public/userA",
	pollingInterval: "500"
}
service<ftp> fileService{
	resource fileResource (message m){
		system:println(messages:getStringPayload(m));
		reply m;
	}
}
```

FTP Client Connector sample
```
import ballerina.lang.system;
import ballerina.net.ftp;
import ballerina.lang.files;
import ballerina.lang.blobs;
import ballerina.lang.strings;

function main(string[] args) {   
	ftp:ClientConnector c = create ftp:ClientConnector();

        // Checking whether a given file exists
	files:File target = {path : "ftp://localhost/public/userA/aa.txt"};
	boolean b = ftp:ClientConnector.exists(c, target);
	system:println("File exists: " + b);

        // Creating a new file/folder 
	files:File newDir = {path : "ftp://localhost/public/userA/new-dir/rec"};
	ftp:ClientConnector.createFile(c, newDir, "folder");

        // Reading a given file. 
	files:File txtFile = {path : "ftp://localhost/public/userA/bb.txt"};
	blob txtb = ftp:ClientConnector.read(c, txtFile);    
	system:println(blobs:toString(txtb, "UTF-8"));

        // Copying a given file to a given location
	files:File copyOfTxt = {path : "ftp://localhost/public/userA/copy.txt"};
	ftp:ClientConnector.copy(c, txtFile, copyOfTxt);

        // Moving a given file to a given location
	files:File mvSrc = {path : "ftp://localhost/public/userA/aa.txt"};
	files:File mvTarget = {path : "ftp://localhost/public/userA/mv/test-move"};
	ftp:ClientConnector.move(c, mvSrc, mvTarget);
 
        // Deleting a given file
	files:File del = {path : "ftp://localhost/public/userA/del-file"};
	ftp:ClientConnector.delete(c, del);

        // Writing to a given file
	files:File wrt = {path : "ftp://localhost/public/userA/write-file"};
	blob content = strings:toBlob("Hello World!", "UTF-8");
	ftp:ClientConnector.write(c, content, wrt);
}
```
Consider a scenario where you want to listen to a particular remote directory, A, and when you come across a file, you want to 
- read the content of the file
- append a new message to that content
- write this new content to a file in directory 'B'
- move the original file to a directory 'C'

The sample given below  demonstrates how to utilize the FTP connectors to address the above problem.
```
import ballerina.net.ftp;
import ballerina.lang.messages;
import ballerina.lang.files;
import ballerina.lang.blobs;
import ballerina.lang.strings;
import ballerina.lang.system;

@ftp:config {
    dirURI: "ftp://127.0.0.1/public/userA/",
    pollingInterval: "500",
    fileSortAscending: "true",
    fileSortAttribute: "name",
    actionAfterProcess: "MOVE",
    moveAfterProcess: "ftp://127.0.0.1/public/userB/",
    parallel: "false"
}
service<ftp> serverCon {
    resource fileResource (message m){
        ftp:ClientConnector c = create ftp:ClientConnector();
        string url = messages:getStringPayload(m);

        files:File file = {path: url};

        system:println(url);

        blob txt = ftp:ClientConnector.read(c, file);
        string content = blobs:toString(txt, "UTF-8");

        system:println(content);

        content = content + "ballerina";
        blob output = strings:toBlob(content, "UTF-8");

        files:File target = {path: "ftp://127.0.0.1/public/userC/cc.txt"};

        ftp:ClientConnector.write(c, output, target);

        reply m;
    }
}
```

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java [hunk 1]
```diff
@@ -44,7 +44,12 @@ public boolean receive(CarbonMessage cMsg, CarbonCallback carbonCallback) throws
             if (log.isDebugEnabled()) {
                 log.debug("ballerina received a request message");
             }
-            ThreadPoolFactory.getInstance().getExecutor().execute(new RequestWorkerThread(cMsg, carbonCallback));
+            if (cMsg.getProperty("SINGLE_THREADED") != null) {
+                ThreadPoolFactory.getInstance().getSingleThreadExecutor((String) cMsg.getProperty("SINGLE_THREADED")).
+                        execute(new RequestWorkerThread(cMsg, carbonCallback));
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java]: This part is related to file server connector.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/Constants.java [hunk 2]
```diff
@@ -27,4 +27,14 @@
     public static final String PROTOCOL_PACKAGE_FILE = "ballerina.net.file";
     public static final String TRANSPORT_PROPERTY_SERVICE_NAME = "TRANSPORT_FILE_SERVICE_NAME";
     public static final String FILE_PACKAGE_NAME = "ballerina.net.file";
+    public static final String FILE_UPDATE = "FILE_UPDATE";
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/Constants.java]: This part is related to file server connector.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/FileResourceDispatcher.java [hunk 3]
```diff
@@ -39,38 +39,37 @@
...
-        }
-        Resource[] resources = service.getResources();
-        if (resources.length != 1) {
-            throw new BallerinaException("A Service of type '" + Constants.PROTOCOL_FILE
-                    + "' has to have only one resource associated to itself. " + "Found " + resources.length
-                    + " resources in Service: " + service.getSymbolName().getName());
-        }
-        return resources[0];
+                                 Context balContext) throws BallerinaException {
+        return null;
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/FileResourceDispatcher.java]: This method in not needed.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/Constants.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    public static final String FILE_SYSTEM_PACKAGE_NAME = "ballerina.net.fs";
+
+    public static final String ANNOTATION_DIR_URI = "dirURI";
+    public static final String ANNOTATION_FILE_PATTERN = "fileNamePattern";
+    public static final String ANNOTATION_POLLING_INTERVAL = "pollingInterval";
+    public static final String ANNOTATION_CRON_EXPRESSION = "cronExpression";
+    public static final String ANNOTATION_ACK_TIMEOUT = "ackTimeOut";
+    public static final String ANNOTATION_FILE_COUNT = "perPollFileCount";
+
+
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/Constants.java]: Unwanted spacing.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/FileSystemServiceDispatcher.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    public String getProtocolPackage() {
+        return Constants.FILE_SYSTEM_PACKAGE_NAME;
+    }
+
+    @Override
+    public ServiceInfo findService(CarbonMessage cMsg, CarbonCallback callback) {
+        Object serviceNameProperty = cMsg.getProperty(Constants.TRANSPORT_PROPERTY_SERVICE_NAME);
+        String serviceName = (serviceNameProperty != null) ? serviceNameProperty.toString() : null;
+        if (serviceName == null) {
+            throw new BallerinaException("Service name is not found with the file input stream.");
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/FileSystemServiceDispatcher.java]: Better error message would be nice

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/ftp/Constants.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    public static final String TRANSPORT_PROPERTY_SERVICE_NAME = "TRANSPORT_FILE_SERVICE_NAME";
+    public static final String FTP_PACKAGE_NAME = "ballerina.net.ftp";
+
+    public static final String ANNOTATION_DIR_URI = "dirURI";
+    public static final String ANNOTATION_FILE_PATTERN = "fileNamePattern";
+    public static final String ANNOTATION_POLLING_INTERVAL = "pollingInterval";
+    public static final String ANNOTATION_CRON_EXPRESSION = "cronExpression";
+    public static final String ANNOTATION_ACK_TIMEOUT = "ackTimeOut";
+    public static final String ANNOTATION_FILE_COUNT = "perPollFileCount";
+
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/ftp/Constants.java]: Unwanted spacing

File: modules/ballerina-native/src/main/ballerina/ballerina/net/file/annotation.bal [hunk 7]
```diff
@@ -2,10 +2,14 @@ package ballerina.net.file;
 
 annotation FileSource attach service {
```
Comment (shafreenAnfar) [inline modules/ballerina-native/src/main/ballerina/ballerina/net/file/annotation.bal]: This is related to file server

File: modules/ballerina-native/src/main/ballerina/ballerina/net/fs/annotation.bal [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package ballerina.net.fs;
+
+import ballerina.doc;
+
+annotation config attach service {
```
Comment (shafreenAnfar) [inline modules/ballerina-native/src/main/ballerina/ballerina/net/fs/annotation.bal]: Lets merge them

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/ftp/AbstractFtpAction.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                }
+            }
+        } catch (InterruptedException e) {
+            logger.debug("Interrupted while waiting for isExists callback", e);
+        }
+        return ((BMessage) callback.getValueRef()).value();
+    }
+
+    //TODO: Refactor this to check for starts with
+    protected boolean validateProtocol(String url) {
```
Comment (shafreenAnfar) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/ftp/AbstractFtpAction.java]: I guess this is not needed.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/ftp/AbstractFtpAction.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            //Getting the sender instance and sending the message.
+            BallerinaConnectorManager.getInstance().getClientConnector(FileConstants.FTP_CONNECTOR_NAME)
+                                     .send(message, callback, propertyMap);
+        } catch (ClientConnectorException e) {
+            throw new BallerinaException(e.getMessage(), e, context);
+        }
+        try {
+            // Wait till Response comes
+            long startTime = System.currentTimeMillis();
+            while (!callback.isResponseArrived()) {
```
Comment (shafreenAnfar) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/ftp/AbstractFtpAction.java]: Do we really need this loop ?

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/FileSystemServiceDispatcher.java (hunk 5 above)
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/FileSystemServiceDispatcher.java]: Ack.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/FileSystemResourceDispatcher.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public ResourceInfo findResource(ServiceInfo service, CarbonMessage cMsg, CarbonCallback callback) throws
+            BallerinaException {
+        if (log.isDebugEnabled()) {
+            log.debug("Starting to find resource in the file service " + service.getName() + " to " +
+                      "deliver the message");
+        }
+        ResourceInfo[] resourceInfoList = service.getResourceInfoList();
+        if (resourceInfoList.length != 1) {
+            throw new BallerinaException("A Service of type '" + Constants.PROTOCOL_FILE_SYSTEM +
```
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/fs/FileSystemResourceDispatcher.java]: Ack.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/ftp/Constants.java (hunk 6 above)
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/ftp/Constants.java]: Ack.

File: modules/ballerina-native/src/main/ballerina/ballerina/net/file/annotation.bal (hunk 7 above)
Comment (pubudu91) [inline modules/ballerina-native/src/main/ballerina/ballerina/net/file/annotation.bal]: Ack. Removed.

File: modules/ballerina-native/src/main/ballerina/ballerina/net/fs/annotation.bal (hunk 8 above)
Comment (pubudu91) [inline modules/ballerina-native/src/main/ballerina/ballerina/net/fs/annotation.bal]: Ack.

File: modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/ftp/AbstractFtpAction.java (hunk 9 above)
Comment (pubudu91) [inline modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/ftp/AbstractFtpAction.java]: Ack.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java (hunk 1 above)
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java]: Ack.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/Constants.java (hunk 2 above)
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/Constants.java]: Ack. Removed

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/FileResourceDispatcher.java (hunk 3 above)
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/dispatchers/file/FileResourceDispatcher.java]: Ack.

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java [hunk 12]
```diff
@@ -44,8 +44,13 @@ public boolean receive(CarbonMessage cMsg, CarbonCallback carbonCallback) throws
             if (log.isDebugEnabled()) {
                 log.debug("ballerina received a request message");
             }
-            ThreadPoolFactory.getInstance().getExecutor().execute(new RequestWorkerThread(cMsg, carbonCallback));
-        } else {
+//            if (cMsg.getProperty("SINGLE_THREADED") != null) {
+//                ThreadPoolFactory.getInstance().getSingleThreadExecutor((String) cMsg.getProperty("SINGLE_THREADED")).
+//                        execute(new RequestWorkerThread(cMsg, carbonCallback));
+//            } else {
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java]: Can we remove these comments

Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java]: @pubudu91

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/threadpool/ThreadPoolFactory.java [hunk 13]
```diff
@@ -51,6 +54,16 @@ public ExecutorService getExecutor() {
...
+    public ExecutorService getSingleThreadExecutor(String serviceName) {
+        if (!singularThreads.containsKey(serviceName)) {
+            ExecutorService singleThreadExecutor =
+                    Executors.newFixedThreadPool(1, new BLangThreadFactory("BLangWorker"));
+            singularThreads.put(serviceName, singleThreadExecutor);
+            return singleThreadExecutor;
+        }
+        return singularThreads.get(serviceName);
+    }
+
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/threadpool/ThreadPoolFactory.java]: @pubudu91 Not needed for these two connectors

File: modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java (hunk 12 above)
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/services/MessageProcessor.java]: Ack.

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/threadpool/ThreadPoolFactory.java (hunk 13 above)
Comment (pubudu91) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/threadpool/ThreadPoolFactory.java]: Ack

**Pull request merged**